src/AggregateFunctions/AggregateFunctionGroupSortedArray.h

```diff
@@ -0,0 +1,248 @@
+                current_max = threshold;
+
+            //insert element into the given position
+            results[cur] = i;
+        }
+    }
+}
+
+template <bool is_plain_column, typename T, bool is_weighted>
+class AggregateFunctionGroupSortedArray : public IAggregateFunctionDataHelper<
+                                              AggregateFunctionGroupSortedArrayData<T, is_weighted>,
+                                              AggregateFunctionGroupSortedArray<is_plain_column, T, is_weighted>>
+{
+protected:
+    using State = AggregateFunctionGroupSortedArrayData<T, is_weighted>;
+    using Base = IAggregateFunctionDataHelper<AggregateFunctionGroupSortedArrayData<T, is_weighted>, AggregateFunctionGroupSortedArray>;
+
+    UInt64 threshold;
+    DataTypePtr & input_data_type;
+    mutable std::mutex mutex;
+
+    static void deserializeAndInsert(StringRef str, IColumn & data_to);
+
+public:
+    AggregateFunctionGroupSortedArray(UInt64 threshold_, const DataTypes & argument_types_, const Array & params)
+        : IAggregateFunctionDataHelper<AggregateFunctionGroupSortedArrayData<T, is_weighted>, AggregateFunctionGroupSortedArray>(
+            argument_types_, params)
+        , threshold(threshold_)
+        , input_data_type(this->argument_types[0])
+    {
+    }
+
+    void create(AggregateDataPtr place) const override
+    {
+        Base::create(place);
+        this->data(place).threshold = threshold;
+    }
+
+    String getName() const override { return is_weighted ? "groupSortedArrayWeighted" : "groupSortedArray"; }
+
+    DataTypePtr getReturnType() const override { return std::make_shared<DataTypeArray>(input_data_type); }
+
+    bool allocatesMemoryInArena() const override
+    {
+        if constexpr (std::is_same_v<T, StringRef>)
+            return true;
+        else
+            return false;
+    }
+
+    void add(AggregateDataPtr __restrict place, const IColumn **columns, size_t row_num, Arena *arena) const override { 
+        State & data = this->data(place);
+        if constexpr (is_weighted)
+        {
+            data.add(readItem<T, is_plain_column>(columns[0], arena, row_num),
+                     readItem<UInt64, false>(columns[1], arena, row_num));
+        }
+        else
+        {
+             data.add(readItem<T, is_plain_column>(columns[0], arena, row_num));
+        }
+    }
+
+    void addBatchSinglePlace(
+        size_t batch_size, AggregateDataPtr place, const IColumn ** columns, Arena * arena, ssize_t if_argument_pos) const override
+    {
+        State & data = this->data(place);
+        if constexpr (is_weighted)
+        {
+            StringRef ref = columns[1]->getRawData();
+            UInt64 values[batch_size];
+            memcpy(values, ref.data, batch_size * sizeof(UInt64));
+            UInt64 * bestRows = new UInt64[this->threshold];
+            size_t num_results;
+
+            //First store the first n elements with the column number
+            if (if_argument_pos >= 0)
```

Comment:
This looks extremely inefficient and not an improvement over topK (in fact because it's bugged for the -If it's worse).

The way I see if you need a function that given an array of UInt64 gives you back the **indexes** of top `threshold` elements and then you access those only.

Reply:
I'm not sure about when 'if_argument_pos' is greater than 0 and the we run this code. But I understand we it is a filter to apply to the rows, we need first and array with all the data to evaluate so honestly speaking I don't know how to do it better.
I took this code from the base class IAggregateFunction::addBatchSinglePlace().